### DISCUSSION TITLE
Simple Market Maker Strategy that odes not work

### DISCUSSION THREAD

**Discussion by nikky78** [kernc/backtesting.py#612] [Q&A]
Hi guys,

I am trying to build a very simple market making bot, but even though orders are filled over time, position remains 0 and P&L stays 0.
```py
class GridBot(Strategy):
    def init(self):
        super().init()

        self.buy(limit=38395, size=0.1)
        self.sell(limit=44584, size=0.1)

        print('Init orders', self.orders)

    def next(self):
        print('position', self.position)
        print('trades', self.trades)
        print('orders', self.orders)
        print()

bt = Backtest(
    df,
    GridBot,
    cash=100_000, commission=.000,
    trade_on_close=True,
    hedging=True,
    margin=1,
)

stats = bt.run()
```

Would you have any idea why this does not work properly? I would expect to have in the end `(44584-38395)*0.1*100_000` $ of P&L.

Thank you,

**Comment by kernc** [comment]
You call [`Strategy.buy()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.buy) and [`Strategy.sell()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.sell) in [`Strategy.next()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.next).

With [`Backtest(..., hedging=True)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest) (and even without), both your orders in `init()` are placed _simultaneously_.

**Reply by nikky78** [reply]
> You call [`Strategy.buy()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.buy) and [`Strategy.sell()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.sell) in [`Strategy.next()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.next).
> 
> With [`Backtest(..., hedging=True)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest) (and even without), both your orders in `init()` are placed _simultaneously_.

I am just trying to set 2 orders on the buy side and on the sell side, and wait for them to be filled. I am not trying to fill any new order after. So naturally, I am expected to have a long position with the buy order is filled, and then again a neutral position when the sell order is filled. Then when both orders are filled, I should see a higher equity. But when printing trades, nothing happens when one order is filled, however, I see the number of open orders reducing over time.

**Reply by kernc** [reply]
> So naturally, I am expecting to have a long position when the buy order is filled, and then again a neutral position when the sell order is filled.

No, because of [`Backtest(..., hedging=True)`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Backtest), a short/sell order _does not_ close a long/buy order. You'll need to use [`trade.close()`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Trade.close) or `Strategy.sell(size=long_trade.size)` with `Backtest(..., hedging=False)`.

**Reply by nikky78** [reply]
> hedging=False

Okay, but here I am setting a buy limit of 10,000$ and a sell limit of 10,000$. So when one price is reached, I should get a position of +/- 10,000$ and when the second price is reached I should get a neutral position with a positif P&L

**Reply by nikky78** [reply]
I really don;t get it. How would you write that simple strategy of just placing one buy one sell limit order at inception, and just let the simulation run?

**Reply by nikky78** [reply]
I get 

40095.0
position Position: 0 (0 trades)
orders (Order size=10000.0, limit=38429.0, contingent=0, Order size=-10000.0, limit=44442.0, contingent=0)

38395.0
position Position: 0 (0 trades)
orders (Order size=-10000.0, limit=44442.0, contingent=0,)

44443.0
position Position: 0 (0 trades)
orders ()

And strategy.trades never show anything too. So it seems that orders are filled when price crosses the limit price, ut still position remains at 0

**Reply by kernc** [reply]
Try this. It opens and closes a single trade. The equity is changed accordingly.
```py
from backtesting import Strategy, Backtest
from backtesting.test import GOOG


class GridBot(Strategy):
    def init(self):
        self.buy(limit=300, size=1)  # Absolute size in "units"
        self.sell(limit=600, size=1)
        print('Init orders', self.orders)

    def next(self):
        print(self.data.index[-1], self.position, self.trades, self.orders)


bt = Backtest(
    GOOG,
    GridBot,
    trade_on_close=True,
    hedging=False,
)
stats = bt.run()
stats
```
